### ISSUE TITLE
Unable to compile on Ubuntu

### ISSUE THREAD

**Issue by Usohatchi** [wacossusca34/glava#13]
When I run make I get:

> In file included from xwin.c:21:0:
> /usr/include/GLFW/glfw3native.h:102:3: error: #error "No context API selected"
>   #error "No context API selected"
>    ^
> Makefile:59: recipe for target 'xwin.o' failed

**Comment by kartik-bhagatwala** [comment]
I also get this error. I am running Ubuntu 16.04. It seems like it's an error thrown by the GLFW library, but I dont know much past that. 
```
sudo make
cc -I glad/include -O2 -march=native -DGLAVA_UNIX  -o xwin.o -c xwin.c
In file included from xwin.c:21:0:
/usr/include/GLFW/glfw3native.h:102:3: error: #error "No context API selected"
  #error "No context API selected"
   ^
Makefile:59: recipe for target 'xwin.o' failed
make: *** [xwin.o] Error 1
```

**Comment by EliTheCoder** [comment]
Confirmed on Ubuntu 16.04 i386

**Comment by Josefan** [comment]
I get past the xwin.o but get this error instead
```

cc -I glad/include -O2 -march=native -DGLAVA_UNIX  -o pulse_input.o -c pulse_input.c
pulse_input.c:6:26: fatal error: pulse/simple.h: No such file or directory
compilation terminated.
Makefile:59: recipe for target 'pulse_input.o' failed
make: *** [pulse_input.o] Error 1
```

**Comment by Godje** [comment]
Confirmed on Ubuntu 17.10 64bit

**Comment by wacossusca34** [comment]
This is because GLava currently needs GLFW 3.2 or higher. Users here are running into issue #7.

I will have to backport to GLFW 3.1 since it seems that is still being used for Ubuntu and other Debian-based distros.

**Comment by EliTheCoder** [comment]
I'm using the apt packages: libglfw3 and libglfw3-dev

**Comment by wacossusca34** [comment]
@EliTheCoder `libglfw3` on Ubuntu provides GLFW 3.1.2-3, but it looks like newer releases (3.2) are available unless you are on Ubuntu Xenial (LTS).

Again, I will backport here, it's just an issue with the older GLFW headers (hopefully today).

**Comment by wacossusca34** [comment]
Fixed in `unstable`.

**Comment by kartik-bhagatwala** [comment]
I just tried out the new version, and I still run into the same error on Ubuntu. However, Ubuntu users with an old version of GLFW can compile the newest version themselves by following these [instructions.](http://www.glfw.org/docs/latest/compile.html)

Also, to fix @Josefan issue, you just need to install the libpulse-dev package from apt.

@wacossusca34 it would also be great if you could add these instructions for Ubuntu users in the install section to avoid more compiling questions.

**Comment by wacossusca34** [comment]
@kartik-bhagatwala Can you make sure `glava --help` shows `1.2 (stable)`? [This commit](https://github.com/wacossusca34/glava/commit/1e2aad05b5b1310c5914d91604228d4b8e8bfa4f) should have fixed the issue. You can also just try passing `-DGLFW_EXPOSE_NATIVE_GLX` to `CFLAGS`.

> Also, to fix @Josefan's issue, you just need to install the libpulse-dev package from apt.

Pulseaudio is started as a requirement, which obviously includes development headers.

**Comment by Usohatchi** [comment]
Tried both master and unstable branches - still getting the same error as first. I'm on Ubuntu 16.0.4.3 LTS.

**Comment by kartik-bhagatwala** [comment]
@wacossusca34 yes, my glava is showing 1.2. I got it compiled it though following my comment above.

`GLava (glava) 1.2 (stable, unix/fhs)
`

**Comment by wacossusca34** [comment]
Alright. I'll investigate this tomorrow.

**Comment by kartik-bhagatwala** [comment]
Also, I'm on GNOME, and the `transparency="xroot"` does not work, the background is just black, I'm not sure if anyone else can replicate this, and whether or not this is a bug. Let me know if I should file another issue for this.

**Comment by wacossusca34** [comment]
@kartik-bhagatwala ensure you're not actually running over an XWayland session, and then create a separate issue. In my own testing, GNOME has worked fine unless Wayland is being used.

**Comment by wacossusca34** [comment]
Okay, _now_ it should be fixed for GLFW 3.1 in `unstable` (and later on the 1.3 release). I made a silly mistake in the original fix, trying to read `GLFW_VERSION_XXX` macros before `#include <glfw3.h>`.

**Comment by kartik-bhagatwala** [comment]
@wacossusca34 I'm pretty sure I'm running X11:
`echo $XDG_SESSION_TYPE
x11
`

**Comment by Usohatchi** [comment]
> cd glad && python -m glad --generator=c --extensions=GL_EXT_framebuffer_multisample,GL_EXT_texture_filter_anisotropic --out-path=.
> [02/07/2018 20:13:02][INFO	][glad   	]: getting 'gl' specification from SVN
> [02/07/2018 20:13:02][INFO	][glad.opener	]: opening: 'https://raw.githubusercontent.com/KhronosGroup/OpenGL-Registry/master/xml/gl.xml'
> [02/07/2018 20:13:14][INFO	][glad   	]: generating 'gl' bindings
> [02/07/2018 20:13:14][INFO	][glad.opener	]: saving: 'https://www.khronos.org/registry/egl/api/KHR/khrplatform.h' to '/home/l05/glava/glad/include/KHR/khrplatform.h'
> [02/07/2018 20:13:17][INFO	][glad   	]: generating 'gl' bindings - done
> cc -I glad/include -O2 -march=native -DGLAVA_UNIX  -o glad.o ./glad/src/glad.c -c
> cc -I glad/include -O2 -march=native -DGLAVA_UNIX  -o glava.o -c glava.c
> cc -I glad/include -O2 -march=native -DGLAVA_UNIX  -o xwin.o -c xwin.c
> xwin.c:27:2: error: #error "GLX defined"
>  #error "GLX defined"
>   ^
> Makefile:59: recipe for target 'xwin.o' failed
> make: *** [xwin.o] Error 1

Compiled for a bit longer this time. Error is the same but seems to appear later in the file.
(xwin.c:27:2 now vs xwin.c:21:0 before)

**Comment by wacossusca34** [comment]
@Usohatchi that was a mistake on my end. It's fixed now.

**Comment by Usohatchi** [comment]
(Sorry, I'm new to github, should I be making more posts instead of tacking on edits? Just gonna copy paste my edit.) 

Removed #error "GLX defined" and xwin.c compiled successfully. Now fails on compiling pulse_input.o. Error I now get:

> pulse_input.c:6:26: fatal error: pulse/simple.h: No such file or directory
> compilation terminated.
> Makefile:59: recipe for target 'pulse_input.o' failed
> make: *** [pulse_input.o] Error 1

**Comment by wacossusca34** [comment]
@Usohatchi You need `libpulse-dev`. Pulseaudio is a compile requirement, as stated in the readme, which means you need development headers for it.

I think I will go with @kartik-bhagatwala's suggestion to add clear instructions for Ubuntu since this seems to be a reoccurring misunderstanding.

**Comment by Usohatchi** [comment]
My bad, installed it and got farther this time, but have ran into some more GLFW-related errors:

> render.c: In function ‘rd_new’:
> render.c:729:35: error: ‘GLFW_FALSE’ undeclared (first use in this function)
>      glfwWindowHint(GLFW_FLOATING, GLFW_FALSE);
>                                    ^
> render.c:729:35: note: each undeclared identifier is reported only once for each function it appears in
> In file included from render.c:21:0:
> render.c: In function ‘_handler’:
> render.c:815:37: error: ‘GLFW_MAXIMIZED’ undeclared (first use in this function)
>          WINDOW_HINT("setmaximized", GLFW_MAXIMIZED),
>                                      ^
> glsl_ext.h:3:53: note: in definition of macro ‘RHANDLER’
>      ({ void _handler(const char* name, void** args) __VA_ARGS__ _handler; })
>                                                      ^
> render.c:815:9: note: in expansion of macro ‘WINDOW_HINT’
>          WINDOW_HINT("setmaximized", GLFW_MAXIMIZED),
>          ^
> Makefile:59: recipe for target 'render.o' failed
> make: *** [render.o] Error 1

**Comment by wacossusca34** [comment]
\*sighs\*

Apparently there are some other incompatibilities with GLFW 3.1, the macros in your error message are not defined. I don't really have an older build of GLFW to play with so I've just been pushing changes as I read the older headers.

**Comment by wacossusca34** [comment]
@Usohatchi please test the recent fix. I just addressed the issues mentioned by the errors you posted.

**Comment by Usohatchi** [comment]
More errors in render.c :/ Hopefully there'll not be too much more.

> In file included from render.c:29:0:
> render.c: In function ‘_handler’:
> render.c:830:37: error: ‘GLFW_MAXIMIZED’ undeclared (first use in this function)
>          WINDOW_HINT("setmaximized", GLFW_MAXIMIZED),
>                                      ^
> glsl_ext.h:3:53: note: in definition of macro ‘RHANDLER’
>      ({ void _handler(const char* name, void** args) __VA_ARGS__ _handler; })
>                                                      ^
> render.c:830:9: note: in expansion of macro ‘WINDOW_HINT’
>          WINDOW_HINT("setmaximized", GLFW_MAXIMIZED),
>          ^
> render.c:830:37: note: each undeclared identifier is reported only once for each function it appears in
>          WINDOW_HINT("setmaximized", GLFW_MAXIMIZED),
>                                      ^
> glsl_ext.h:3:53: note: in definition of macro ‘RHANDLER’
>      ({ void _handler(const char* name, void** args) __VA_ARGS__ _handler; })
>                                                      ^
> render.c:830:9: note: in expansion of macro ‘WINDOW_HINT’
>          WINDOW_HINT("setmaximized", GLFW_MAXIMIZED),
>          ^
> Makefile:59: recipe for target 'render.o' failed
> make: *** [render.o] Error 1

I'm out for the night, I can test some more tomorrow.

**Comment by wacossusca34** [comment]
Should work now with [this commit](https://github.com/wacossusca34/glava/commit/c8cce220d83a07d27932a6b2ff31863ea712878a).

**Comment by Usohatchi** [comment]
Works! Awesome, thanks.
